src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -1544,16 +1544,51 @@ static int addr_exchange_root_vni(MPIR_Comm * init_comm)
     goto fn_exit;
 }
 
+/* NOTE:
+ * Each nic-vni is an endpoint with a unique address, and inside libfabric maintains
+ * one av table. Thus to fully store the address mapping, we'll need a multi-dim table as
+ *     av_table[src_nic][src_vni][dest_rank][dest_nic][dest_vni]
+ * Note, this table is for illustration, and different from MPIDI_OFI_addr_t.
+ *
+ * However, if we insert the address carefully, we can manage to make the av table inside
+ * each endpoint *indentical*. Then, we can omit the dimension of [src_nic][src_vni].
+ *
+ * To achieve that, we use the following 3-step process (described with above illustrative av_table).
+ *
+ * 1. insert and store       av_table[ 0 ][ 0 ][rank][ 0 ][ 0 ]
+ *
+ * 2. insert and store       av_table[ 0 ][ 0 ][rank][nic][vni]
+ *
+ * 3. insert (but not store) av_table[nic][vni][rank][nic][vni]
+ *
+ * The step 1 is done in addr_exchange_root_vni. Step 2 and 3 are done in addr_exchange_all_vnis.
+ * Step 3 populates av tables inside libfabric for all non-zero endpoints, but they should be
+ * identical to the table in root endpoint, thus no need to store them in mpich. Thus the table is
+ * reduced to
+ *      av_table[rank] -> dest[nic][vni]
+ */
```

Comment:
@wesbland This is what I regurgitated out after digesting your explanations. What do you think?

Reply:
Yeah, I think this is fine.